File: dec4/two.mjs
import fs from "fs/promises";

const STRING = "MAS";
const directions = [
    [-1, -1], // up-left
    [-1, 1],  // up-right
    [1, -1],  // down-left
    [1, 1],   // down-right
];

(async function () {
    let file = await fs.readFile("./input.txt", "utf-8");
    let text = file.trim().split("\n");
    let count = 0;
    for(let i = 0; i < text.length; i++) {
        for(let j = 0; j < text[i].length; j++) {
            if(text[i][j] === "M") {
                for(let dir of directions) {
                    if(find(text, j, i, dir, 0)) {
                        if(find(text, j, i + 2 * dir[0], [-dir[0], dir[1]], 0)) {
                            count++;
                        }
                        else if(find(text, j + 2 * dir[1], i, [dir[0], -dir[1]], 0)) {
                            count++;
                        }
                    }
                }
            }
        }
    }
    console.log(count / 2);
})()

function find (text, x, y, dir, index) {
    if(index === STRING.length) return true;
    if(y < 0 || y >= text.length || x < 0 || x >= text[y].length) return false;
    if(text[y][x] !== STRING[index]) return false;
    return find(text, x + dir[1], y + dir[0], dir, index + 1);
}

File: dec4/two.rs
use std::fs;

fn main () -> Result<(), Box<dyn std::error::Error>> {
    let file = fs::read_to_string("./input.txt")?;
    let mut text = file.trim().lines();
    let mut count = 0;
    let directions: [(i32, i32); 4] = [(-1, -1), (-1, 1), (1, -1), (1, 1)];
    
    for (i, line) in text.clone().enumerate() {
        for (j, character) in line.chars().enumerate() {
            if character == 'M' {
                for (dx, dy) in directions.iter() {
                    if find(&mut text, dx, dy, j, i, 0) {
                        if find(&mut text, &-dx, dy, ((j as i32) + 2 * dx) as usize, i, 0) {
                            count += 1;
                        }
                        if find(&mut text, dx, &-dy, j, ((i as i32) + 2 * dy) as usize, 0) {
                            count += 1;
                        }
                    }
                }
            }
        }
    }
    
    println!("{}", count / 2);
    Ok(())
}

fn find(text: &mut std::str::Lines, dx: &i32, dy: &i32, x: usize, y: usize, i: i32) -> bool {
    let binding = "MAS".to_string();
    let string = binding.chars();
    if i == string.clone().count() as i32 {
        return true;
    }
    let x = x as i32;
    let y = y as i32;
    if x < 0 || y < 0 || x >= text.clone().nth(0).unwrap().len() as i32 || y >= text.clone().collect::<Vec<_>>().len() as i32 {
        return false;
    }
    if text.clone().nth(y as usize).unwrap().as_bytes()[x as usize] as char != string.clone().nth(i as usize).unwrap() {
        return false;
    }
    return find(text, dx, dy, (x + dx) as usize, (y + dy) as usize, i + 1);
}